src/wasm/wasm-type.cpp

```diff
@@ -3241,15 +3263,24 @@ std::vector<HeapType> TypeBuilder::build() {
   }
 
   // Combine the basic types with the built types.
-  std::vector<HeapType> result(entryCount);
+  std::vector<HeapType> results(entryCount);
   for (size_t i = 0, builtIndex = 0; i < entryCount; ++i) {
     if (basicHeapTypes[i]) {
-      result[i] = *basicHeapTypes[i];
+      results[i] = *basicHeapTypes[i];
     } else {
-      result[i] = built[builtIndex++];
+      results[i] = built[builtIndex++];
     }
   }
-  return result;
+
+  // Note built signature types. See comment in `HeapType::HeapType(Signature)`.
+  for (auto type : results) {
+    if (type.isSignature() &&
+        (type.isNominal() || getTypeSystem() == TypeSystem::Nominal)) {
+      nominalSignatureCache.insertType(type);
+    }
+  }
```

Comment:
Why is this needed when we have it handled in the constructor of HeapType?

Reply:
`TypeBuilder` only uses the lower-level constructor that takes a `uintptr_t` pointing to an underlying `HeapTypeInfo`, so that logic in the other constructor does not otherwise get a chance to run.

Reply:
Thanks, right, that makes sense...